src/v/storage/segment_deduplication_utils.cc

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "storage/segment_deduplication_utils.h"
+
+#include "storage/compacted_index_writer.h"
+#include "storage/compaction_reducers.h"
+#include "storage/index_state.h"
+#include "storage/key_offset_map.h"
+#include "storage/probe.h"
+#include "storage/segment.h"
+#include "storage/segment_set.h"
+#include "storage/segment_utils.h"
+#include "storage/types.h"
+
+namespace storage {
+
+namespace {
+ss::future<ss::stop_iteration> put_entry(
+  key_offset_map& map,
+  compacted_index::entry&& idx_entry,
+  bool& fully_indexed) {
+    auto offset = idx_entry.offset + model::offset_delta(idx_entry.delta);
+    bool success = co_await map.put(idx_entry.key, offset);
+    if (success) {
+        co_return ss::stop_iteration::no;
+    }
+    fully_indexed = false;
+    co_return ss::stop_iteration::yes;
+}
+
+ss::future<bool> should_keep(
+  const key_offset_map& map,
+  const model::record_batch& b,
+  const model::record& r) {
+    const auto o = b.base_offset() + model::offset_delta(r.offset_delta());
+    auto key_view = compaction_key{iobuf_to_bytes(r.key())};
+    auto key = enhance_key(
+      b.header().type, b.header().attrs.is_control(), key_view);
+    auto map_offset = co_await map.get(key);
+    // If the map hasn't indexed the given key, we should keep the
+    // key.
+    if (!map_offset.has_value()) {
+        co_return true;
+    }
+    // We should only keep the record if its offset is equal or
+    // higher than that indexed.
+    co_return map_offset.value() <= o;
```

Comment:
Can the offset ever be equal to the indexed one?

Reply:
Yeah, the offset being equal means that this record was the highest indexed offset for the key.

The offset being higher means that the record has a higher offset than what we've indexed, e.g. because we only partially indexed a segment and a later offset in the segment had the given key at a higher offset.

Reply:
Right. We iterate through the segments in reverse order, but read from the compaction index in the natural order.

Reply:
Right, exactly